### PR TITLE
VSI Rewrite: Add API endpoints

### PR DESCRIPTION
**Pull request by kitified**
# Overview

Adds functions to use VSI API endpoints.

## Acceptance criteria

Feature parity with VSI

## Testing plan

Unused, so not really tested. These are 1:1 with the implementations in the VSI plugin though, so unlikely to require significant change.

## Risks

Nothing really risky, especially since they're currently unused.

## References

Closes https://github.com/fossas/team-analysis/issues/843

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I updated `*schema.json` if I have made changes for `.fossa.yml`, `fossa-deps.{json, yaml, yml}`. You may also need to update these if you have added/removed new dependency (e.g. pip) or analysis target type (e.g. poetry).
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

File: src/App/Fossa/FossaAPIV1.hs [hunk 1]
```diff
@@ -580,3 +587,100 @@ resolveProjectDependencies apiOpts locator = fossaReq $ do
...
+    (status :: Text) <- obj .: "Status"
+    pure . VSIScanAnalysisStatusBody $ VSI.parseAnalysisStatus status
+
+vsiScanAnalysisStatusEndpoint :: Url scheme -> VSI.ScanID -> Url scheme
+vsiScanAnalysisStatusEndpoint baseurl scanID = baseVsiUrl baseurl /: "scans" /: toText scanID /: "status" /: "analysis"
+
+vsiScanAnalysisStatus :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> VSI.ScanID -> m VSI.AnalysisStatus
+vsiScanAnalysisStatus apiOpts scanID = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+  (resBody :: VSIScanAnalysisStatusBody) <- responseBody <$> req GET (vsiScanAnalysisStatusEndpoint baseUrl scanID) NoReqBody jsonResponse baseOpts
```
Comment (scruffystuffs) [inline src/App/Fossa/FossaAPIV1.hs:669]: Is this type annotation needed?  I think the newtype unwrapping function below hints the type well enough.

File: src/App/Fossa/VSI/Types.hs [hunk 2]
```diff
@@ -24,6 +27,27 @@ import Effect.Logger (Pretty (pretty), viaShow)
 import Srclib.Converter (depTypeToFetcher, fetcherToDepType)
 import Srclib.Types qualified as Srclib
 
+-- | The VSI backend returns a scan ID when a scan is created, which is then used to add files to the scan and get inferred OSS dependencies.
+newtype ScanID = ScanID {unScanID :: Text} deriving (ToJSON, FromJSON)
+
+instance ToText ScanID where
+  toText = unScanID
```
Comment (scruffystuffs) [inline src/App/Fossa/VSI/Types.hs:34]: I think we should avoid creating `ToText` instances for `Text`-wrapping newtypes.  It kind of breaks the semantic safety that they offer.  See my comment on `vsiAddFilesToScanEndpoint` for an example.

File: src/App/Fossa/FossaAPIV1.hs [hunk 3]
```diff
@@ -580,3 +587,100 @@ resolveProjectDependencies apiOpts locator = fossaReq $ do
...
+  (resBody :: VSICreateScanResponseBody) <- responseBody <$> req POST (vsiCreateScanEndpoint baseUrl) (ReqBodyJson reqBody) jsonResponse baseOpts
+  pure $ unVSICreateScanResponseBody resBody
+
+newtype VSIAddFilesToScanRequestBody = VSIAddFilesToScanRequestBody {vsiAddFilesToScanRequestBodyFiles :: Map (Path Rel File) Fingerprint.Combined}
+
+instance ToJSON VSIAddFilesToScanRequestBody where
+  toJSON VSIAddFilesToScanRequestBody{..} = object ["ScanData" .= vsiAddFilesToScanRequestBodyFiles]
+
+vsiAddFilesToScanEndpoint :: Url scheme -> VSI.ScanID -> Url scheme
+vsiAddFilesToScanEndpoint baseurl scanID = baseVsiUrl baseurl /: "scans" /: toText scanID /: "files"
```
Comment (scruffystuffs) [inline src/App/Fossa/FossaAPIV1.hs:635]: Using a `toText` here kind of breaks the type safety that we gain with the `ScanID` newtype.
I would usually prefer to see something like this:
```suggestion
vsiAddFilesToScanEndpoint baseurl (VSI.ScanID scanID) = baseVsiUrl baseurl /: "scans" /: scanID /: "files"
```

File: src/App/Fossa/FossaAPIV1.hs [hunk 4]
```diff
@@ -580,3 +587,100 @@ resolveProjectDependencies apiOpts locator = fossaReq $ do
...
+
+vsiCreateScan :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> ProjectRevision -> m VSI.ScanID
+vsiCreateScan apiOpts ProjectRevision{..} = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+
+  Organization orgId _ <- getOrganization apiOpts
+  let projectID = renderLocator $ Locator "custom" projectName Nothing
+  let reqBody = VSICreateScanRequestBody orgId projectRevision projectID
+
+  (resBody :: VSICreateScanResponseBody) <- responseBody <$> req POST (vsiCreateScanEndpoint baseUrl) (ReqBodyJson reqBody) jsonResponse baseOpts
```
Comment (scruffystuffs) [inline src/App/Fossa/FossaAPIV1.hs:626]: Is this type annotation needed?  I think the newtype unwrapping function below hints the type well enough.

**Review by scruffystuffs — APPROVED** [review]
Left a few minor comments, nothing blocking.

Also, I didn't check that the URL's were actually the correct ones, since I'm not sure what the correct ones are.  If we're able to integration test this at some point, that would be a good idea, but I think we don't have a mechanism for that today.

File: src/App/Fossa/FossaAPIV1.hs (hunk 1 above)
Comment (kitified) [reply src/App/Fossa/FossaAPIV1.hs:669]: Oh! I didn't even think of that. Nice point!

File: src/App/Fossa/FossaAPIV1.hs (hunk 3 above)
Comment (kitified) [reply src/App/Fossa/FossaAPIV1.hs]: Fair enough, I buy that argument. Thanks!

**Pull request merged**